sw/host/tests/chip/gpio/src/sleep_pin_mio_dio_val.rs

```diff
@@ -0,0 +1,252 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{ensure, Result};
+use clap::Parser;
+use std::time::Duration;
+
+use opentitanlib::app::TransportWrapper;
+use opentitanlib::execute_test;
+use opentitanlib::io::gpio::PinMode;
+use opentitanlib::io::gpio::PullMode;
+use opentitanlib::test_utils::init::InitializeTest;
+use opentitanlib::uart::console::UartConsole;
+
+const MIO_PADS: &[&str] = &[
+    "IOA0",  // MIO Pad 0
+    "IOA1",  // MIO Pad 1
+    "IOA2",  // MIO Pad 2
+    "IOA3",  // MIO Pad 3
+    "IOA4",  // MIO Pad 4
+    "IOA5",  // MIO Pad 5
+    "IOA6",  // MIO Pad 6
+    "IOA7",  // MIO Pad 7
+    "IOA8",  // MIO Pad 8
+    "IOB0",  // MIO Pad 9
+    "IOB1",  // MIO Pad 10
+    "IOB2",  // MIO Pad 11
+    "IOB3",  // MIO Pad 12
+    "IOB4",  // MIO Pad 13
+    "IOB5",  // MIO Pad 14
+    "IOB6",  // MIO Pad 15
+    "IOB7",  // MIO Pad 16
+    "IOB8",  // MIO Pad 17
+    "IOB9",  // MIO Pad 18
+    "IOB10", // MIO Pad 19
+    "IOB11", // MIO Pad 20
+    "IOB12", // MIO Pad 21
+    "IOC0",  // MIO Pad 22
+    "IOC1",  // MIO Pad 23
+    "IOC2",  // MIO Pad 24
+    "IOC3",  // MIO Pad 25
+    "IOC4",  // MIO Pad 26
+    "IOC5",  // MIO Pad 27
+    "IOC6",  // MIO Pad 28
+    "IOC7",  // MIO Pad 29
+    "IOC8",  // MIO Pad 30
+    "IOC9",  // MIO Pad 31
+    "IOC10", // MIO Pad 32
+    "IOC11", // MIO Pad 33
+    "IOC12", // MIO Pad 34
+    "IOR0",  // MIO Pad 35
+    "IOR1",  // MIO Pad 36
+    "IOR2",  // MIO Pad 37
+    "IOR3",  // MIO Pad 38
+    "IOR4",  // MIO Pad 39
+    "IOR5",  // MIO Pad 40
+    "IOR6",  // MIO Pad 41
+    "IOR7",  // MIO Pad 42
+    "IOR10", // MIO Pad 43
+    "IOR11", // MIO Pad 44
+    "IOR12", // MIO Pad 45
+    "IOR13", // MIO Pad 46
+];
+
+// IOR6/7, IOR10-13
+const VALID_MIO_PADS_FPGA_CW310: &[bool] = &[
```

Comment:
I would be a little worried about these arrays getting out of sync. What do you think of laying them out in structs? e.g.

```rust
struct Pad {
    name: &'static str,
    valid_cw310: bool,
    valid_silicon: bool,
}

const PADS: &[Pad] = &[
    Pad { name: "IOA0", valid_cw310: false, valid_silicon: true },
    Pad { name: "IOA1", valid_cw310: false, valid_silicon: true },
    // ...
];
```

Reply:
I think the better solution would be to have a proper description of the pins somewhere in the tree (maybe a json file?) that opentitanlib can read. It can be more subtle than just valid or not, some pins have pull-up/down or are connected to peripheral which can make them unusuable for some tests but maybe not all.

Reply:
Good point, is that what the `sw/host/opentitanlib/src/app/config/` JSON files are intended to be used for?

Reply:
Those files serve a different purpose at the moment but it feels like the right place to add more information indeed.

Reply:
After @a-will's suggestion [here](https://github.com/lowRISC/opentitan/pull/23275#issuecomment-2128307590) to go with a device tree, for now I've only incorporated @jwnrt's suggestion of the `struct Pad`. That should resolve this discussion, but please check if it.